src/main/java/com/aws/greengrass/mqttclient/MqttClient.java

```diff
@@ -452,29 +453,38 @@ public synchronized void unsubscribe(UnsubscribeRequest request)
         return CompletableFuture.completedFuture(0);
     }
 
-    @SuppressWarnings("PMD.AvoidCatchingThrowable")
+    @SuppressWarnings({"PMD.AvoidCatchingThrowable", "PMD.PreserveStackTrace"})
     protected CompletableFuture<Integer> publishSingleSpoolerMessage() throws InterruptedException {
-        final long id = spool.popId();
-        PublishRequest request = spool.getMessageById(id);
-        if (request == null) {
-            return null;
-        }
-        MqttMessage m = new MqttMessage(request.getTopic(), request.getPayload());
+        long id = -1L;
         try {
+            id = spool.popId();
+            PublishRequest request = spool.getMessageById(id);
+            MqttMessage m = new MqttMessage(request.getTopic(), request.getPayload());
+
+            long finalId = id;
             return getConnection(false).publish(m, request.getQos(), request.isRetain())
                     .whenComplete((packetId, throwable) -> {
                         // packetId is the SDK assigned ID. Ignore this and instead use the spooler ID
                         if (throwable == null) {
-                            spool.removeMessageById(id);
-                            logger.atDebug().kv("id", id).kv("topic", request.getTopic())
+                            spool.removeMessageById(finalId);
+                            logger.atDebug().kv("id", finalId).kv("topic", request.getTopic())
                                     .log("Successfully published message");
                         } else {
-                            spool.addId(id);
-                            logger.atError().log("Failed to publish the message via Spooler", throwable);
+                            spool.addId(finalId);
+                            logger.atError().log("Failed to publish the message via Spooler and will retry",
+                                    throwable);
                         }
                     });
         } catch (Throwable t) {
-            spool.addId(id);
+            // valid id is starting from 0
+            if (id >= 0) {
+                spool.addId(id);
+            }
+
+            if (Utils.getUltimateCause(t) instanceof InterruptedException) {
+                throw new InterruptedException("Interrupted while publishing from spooler");
+            }
```

Comment:
this isn't necessary, we already check the cause in the caller of this method by looking at the cause of the execution exception. Did you have any additional reasoning that I'm not getting?

Reply:
In the future, the maxInFlightPublishes may set up bigger than one. Thus, the exception would not be necessary to be thrown out and catch by the 'runSpooler' https://github.com/aws-greengrass/aws-greengrass-nucleus/blob/main/src/main/java/com/aws/greengrass/mqttclient/MqttClient.java#L472

Reply:
Even in that case though I believe it would work since we call `.get()` on the combined future which would throw up the exception since that future would be completed already. I suppose this code can't hurt, just think it isn't really necessary.

Reply:
I mean if that interruptedException happen, we can get it immediately. Just a little bitter concern whether it is possible that the message is still publishing, even thought the Mqttclient already requested to be shut down.